src/core/MIGRATION_EXAMPLES.md

```diff
@@ -1003,4 +1007,178 @@ setup(core: CoreSetup){
+const { body } = await client.asInternalUser.get({ id: 'id' });
+```
+
+- the returned error types changed 
+
+There are no longer specific errors for every HTTP status code (such as `BadRequest` or `NotFound`). A generic
+`ResponseError` with the specific `statusCode` is thrown instead.
+
+```ts
+// before
+import { errors } from 'elasticsearch';
+try {
+  await legacyClient.callAsInternalUser('ping');
+} catch(e) {
+  if(e instanceof errors.NotFound) {
+    // do something
+  }
+}
+``` 
+
+```ts
+// after
+import { errors } from '@elastic/elasticsearch';
+try {
+  await client.asInternalUser.ping();
+} catch(e) {
+  if(e instanceof errors.ResponseError && e.statusCode === 404) {
+    // do something
+  }
+  // also possible, as all errors got a name property with the name of the class,
+  // so this slightly better in term of performances
+  if(e.name === 'ResponseError' && e.statusCode === 404) {
+    // do something
+  }
+}
+```
+
+Please refer to the  [Breaking changes list](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/breaking-changes.html)
+for more information about the changes between the legacy and new client.
+
+### Accessing the client from a route handler
+
+Apart from the API format change, accessing the client from within a route handler
+did not change. As it was done for the legacy client, a preconfigured scoped client 
+bound to the request is accessible using `core` context provider:
+
+```ts
+// before
+router.get(
+  {
+    path: '/my-route',
+  },
+  async (context, req, res) => {
+    const { client } = context.core.elasticsearch.legacy;
+    // call as current user
+    const res = await client.callAsCurrentUser('ping');
+    // call as internal user
+    const res2 = await client.callAsInternalUser('search', options);
+    return res.ok({ body: 'ok' });
+  }
+);
+```
+
+```ts
+// after
+router.get(
+  {
+    path: '/my-route',
+  },
+  async (context, req, res) => {
+    const { client } = context.core.elasticsearch;
+    // call as current user
+    const res = await client.asCurrentUser.ping();
+    // call as internal user
+    const res2 = await client.asInternalUser.search(options);
+    return res.ok({ body: 'ok' });
+  }
+);
+```
+
+### Creating a custom client
+
+The API to create custom clients did not change much:
+
+```ts
+// before
+const customClient = coreStart.elasticsearch.legacy.createClient('my-custom-client', customConfig);
+// do something with the client, such as
+await customClient.callAsInternalUser('ping');
+// custom client are closable
+customClient.close();
+```
+
+```ts
+// after
+const customClient = coreStart.elasticsearch.createClient('my-custom-client', customConfig);
+// do something with the client, such as
+await customClient.asInternalUser.ping();
+// custom client are closable
+customClient.close();
+```
+
+Note that the `plugins` option is now longer available on the new client. As the API is now exhaustive, adding custom
```

Comment:
I think we almost always created custom clients to extend the API. Maybe this paragraph should go to the top of this section as most people will probably conclude that they can just remove their custom client (or use `request.transport`. Then if someone still needs a custom client they can read further.

Reply:
Valid point. AFAIK this is the only usage of custom clients